### PR TITLE
[Material] Add support for customizing active + disabled state color for selection controls.

### PR DESCRIPTION
**Pull request by johnsonmh**
## Description

This PR adds the a `MaterialStateProperty<Color>` parameter to  `Checkbox`, `Switch`, and `Radio`.

This allows for things like customizing the _active + enabled_ state of selection controls:

||Before|After|
|---|---|---|
|Switches|Screen Shot 2020-10-28 at 8 19 16 PM|Screen Shot 2020-11-05 at 12 30 50 PM|
|Checkboxes|Screen Shot 2020-10-22 at 7 16 39 PM|Screen Shot 2020-10-22 at 7 16 11 PM|
|Radio|Screen Shot 2020-10-28 at 8 19 04 PM|Screen Shot 2020-10-28 at 8 24 11 PM|

The "After" is accomplished with the following snippet:

```dart
Color? getActiveColor(Set<MaterialState> states) {
  if (states.contains(MaterialState.selected)) {
    final double opacity = states.contains(MaterialState.disabled) ? .38 : 1.0;
    return Colors.purple.withOpacity(opacity);
  }
  return null; // defer to the default
}

Checkbox(
  fillColor: MaterialStateProperty.resolveWith<Color?>(getActiveColor),
  ...
),
```

## Tests

I added the following tests:

* For `Switch`, `Checkbox`, and `Radio`
  * `MaterialStateProperty` parameter works in enabled, disabled, focused and hovered states.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/switch.dart [hunk 1]
```diff
@@ -758,13 +808,17 @@ class _RenderSwitch extends RenderToggleable {
...
       : inactiveTrackColor;
 
-    final Color thumbColor = isEnabled
+    final Color lerpedThumbColor = isEnabled || useActiveColorInDisabledState
       ? Color.lerp(inactiveColor, activeColor, currentValue)!
       : inactiveColor;
+    // Blend the thumb color against a `surfaceColor` background in case the
+    // thumbColor is not opaque. This way we do not see through the thumb to the
+    // track underneath.
+    final Color thumbColor = Color.alphaBlend(lerpedThumbColor, surfaceColor);
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/switch.dart:821]: tests look great and thorough, but I couldn't find a place where this was tested. Can you add an expect or a comment where applicable?

File: packages/flutter/lib/src/material/checkbox.dart [hunk 2]
```diff
@@ -268,8 +286,9 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin {
           return _CheckboxRenderObjectWidget(
             value: widget.value,
             tristate: widget.tristate,
-            activeColor: widget.activeColor ?? themeData!.toggleableActiveColor,
+            activeColor: effectiveActiveColor,
             checkColor: widget.checkColor ?? const Color(0xFFFFFFFF),
+            useActiveColorInDisabledState: activeColor is MaterialStateProperty<Color>,
             inactiveColor: enabled ? themeData!.unselectedWidgetColor : themeData!.disabledColor,
             focusColor: widget.focusColor ?? themeData.focusColor,
```
Comment (perclasson) [inline packages/flutter/lib/src/material/checkbox.dart:293]: Instead of adding `useActiveColorInDisabledState`, can't you just update the `inactiveColor` when it's not enabled? Maybe I am missing something here.
```dart
final inactiveColor;
if (enabled) {
  inactiveColor = themeData!.unselectedWidgetColor;
else {
  inactiveColor = activeColor is MaterialStateProperty<Color> ? effectiveActiveColor : themeData!.disabledColor;
}
```

I assume those lerp functions wouldn't be used anyways since it can't change state when disabled?

Comment (rami-a) [reply packages/flutter/lib/src/material/checkbox.dart:293]: Yeah I feel like you can do something like this and avoid adding the new bool

File: packages/flutter/lib/src/material/switch.dart (hunk 1 above)
Comment (rami-a) [inline packages/flutter/lib/src/material/switch.dart:821]: Is it possible that this is a breaking change?

**Review by rami-a — APPROVED** [review]

File: packages/flutter/lib/src/material/checkbox.dart [hunk 3]
```diff
@@ -389,6 +413,8 @@ class _RenderCheckbox extends RenderToggleable {
          hovering: hovering,
        );
 
+  bool useActiveColorInDisabledState;
```
[the remark this answers is not in the document]
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:416]: Good call, I added some dart docs explaining why we need this bool is needed.

I'd prefer not to add it to the RenderToggleable, since it is only needed for the implementation of Switches, Radio buttons, and checkboxes, and adding it to RenderToggleable would make it public API.

File: packages/flutter/lib/src/material/radio.dart [hunk 4]
```diff
@@ -239,6 +239,17 @@ class Radio<T> extends StatefulWidget {
...
+  ///
+  /// If [activeColor] is a [MaterialStateColor], it will be resolved in the
+  /// following states:
+  ///
+  ///  * [MaterialState.selected].
+  ///  * [MaterialState.hovered].
+  ///  * [MaterialState.focused].
+  ///  * [MaterialState.disabled].
+  ///
+  /// Note, when this [Radio] is not selected, the [activeColor] color will
```
[the remark this answers is not in the document]
Comment (johnsonmh) [reply packages/flutter/lib/src/material/radio.dart:251]: Good suggestion, thanks.

File: packages/flutter/lib/src/material/switch.dart (hunk 1 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/switch.dart:821]: I think it technically could be, but I don't believe it is considered one because none of the customer/internal tests fail. 

It would only break if someone was passing in a non-opaque value to the thumb color. I don't think people do that because 1) Before now, you couldn't customize disabled color (which is when opacity is typically used), and 2) Before now, if you passed in a non-opaque color, you would see the track underneath, which is probably not preferred.

Comment (johnsonmh) [reply packages/flutter/lib/src/material/switch.dart:823]: Good call, added a test.

File: packages/flutter/lib/src/material/checkbox.dart (hunk 2 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:298]: We discussed this offline, I definitely would prefer do something along these lines because I don't prefer the bool I added. I'm only hesitant because using `activeColor` in the `inactive` state would be a little misleading. If it was called "backgroundColor", I would be more comfortable applying it in the inactive state, and overriding `inactiveColor`. 

For now I think I will leave my solution as is as it is the simplest change to the public facing API at the expense of some plumbing in the implementation.

File: packages/flutter/lib/src/material/checkbox.dart [hunk 5]
```diff
@@ -253,12 +270,12 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin {
...
       widget.mouseCursor ?? MaterialStateMouseCursor.clickable,
-      <MaterialState>{
-        if (!enabled) MaterialState.disabled,
-        if (_hovering) MaterialState.hovered,
-        if (_focused) MaterialState.focused,
-        if (widget.tristate || widget.value!) MaterialState.selected,
-      },
+      _states,
+    );
+    final Color activeColor = widget.activeColor ?? themeData!.toggleableActiveColor;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:275]: This isn't quite right. If widget.activeColor is non-null, it can still resolve to null. If that happens then the theme color should be used.

File: packages/flutter/lib/src/material/checkbox.dart [hunk 6]
```diff
@@ -131,16 +132,20 @@ class Checkbox extends StatefulWidget {
...
   ///
-  /// If [activeColor] is a [MaterialStateColor], it will be resolved in the
-  /// following states:
+  /// If [fillColor] is non-null, this will be ignored.
+  final Color? activeColor;
+
+  /// The color that fills the checkbox when it is checked, in all
+  /// [MaterialState]s.
+  ///
+  /// If this is provided, it will be used over [activeColor].
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:141]: And hoverColor, focusColor, right?

File: packages/flutter/lib/src/material/checkbox.dart [hunk 7]
```diff
@@ -63,6 +63,7 @@ class Checkbox extends StatefulWidget {
     required this.onChanged,
     this.mouseCursor,
     this.activeColor,
+    this.fillColor,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:66]: We should assert that only fillColor or activeColor, focusColor, hoverColor are specified.

**Review by HansMuller — COMMENTED** [review]
Looks like this PR needs to be sync'd

File: packages/flutter/lib/src/material/checkbox.dart (hunk 6 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:141]: No, because `hoverColor` and `focusColor` describe the overlay color, `fillColor` is just for the painted checkbox itself.

File: packages/flutter/lib/src/material/checkbox.dart (hunk 7 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:66]: Good call, added asserts to checkbox and radio (already had them on Switch).

File: packages/flutter/lib/src/material/checkbox.dart [hunk 8]
```diff
@@ -236,6 +252,38 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin {
...
+      }
+      return null;
+    });
+  }
+
+  MaterialStateProperty<Color> get _defaultFillColor {
+    final ThemeData themeData = Theme.of(context);
+    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
+      if (states.contains(MaterialState.disabled)) {
+        return themeData.disabledColor;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:278]: I suppose it's a separate PR bug: shouldn't this color and the unselectedWidgeColor be replaced by ColorScheme colors?

File: packages/flutter/lib/src/material/radio.dart [hunk 9]
```diff
@@ -240,8 +241,23 @@ class Radio<T> extends StatefulWidget {
   /// The color to use when this radio button is selected.
   ///
   /// Defaults to [ThemeData.toggleableActiveColor].
+  ///
+  /// If [fillColor] returns a non-null color in the [MaterialState.selected]
+  /// state, it will be used over this.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/radio.dart:246]: Using "over this" here and elsewhere may not be clear to all readers. It might be clearer to say "instead of this color".

File: packages/flutter/lib/src/material/checkbox.dart [hunk 10]
```diff
@@ -130,8 +131,23 @@ class Checkbox extends StatefulWidget {
...
+  /// [MaterialState]s.
+  ///
+  /// If this is provided, it will be used over [activeColor].
+  ///
+  /// Resolves in the following states:
+  ///  * [MaterialState.selected].
+  ///  * [MaterialState.hovered].
+  ///  * [MaterialState.focused].
+  ///  * [MaterialState.disabled].
+  final MaterialStateProperty<Color?>? fillColor;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/checkbox.dart:149]: It would be helpful to provide a class APIDoc example that overrode fillColor. Likewise for Switch. Could be a separate PR.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/radio.dart (hunk 9 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/radio.dart:246]: Done.

File: packages/flutter/lib/src/material/checkbox.dart (hunk 8 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:278]: Yes, I believe all the selection controls should be updated to default to the color scheme values. I agree that should be a separate (potentially breaking change) pr.

File: packages/flutter/lib/src/material/checkbox.dart (hunk 10 above)
Comment (johnsonmh) [reply packages/flutter/lib/src/material/checkbox.dart:149]: Yes that's a good idea, I will do a followup PR with API docs.
